### PR TITLE
Initialize read fields created during field versioning for non-solvers

### LINKED ISSUE #441: fix field versioning for non solvers

### PR DESCRIPTION
**Pull request by jdahm**
## Technical Description

Adds a pass adding multistages that initialize versioned fields that are read before written. Additionally, it fixes spelling errors in comments, and adds versioned name to the stencil instantiation output.

### Resolves / Enhances

Resolves #441.

### Example

This resolves the simple case:
```
#include "gridtools/clang_dsl.hpp"
using namespace gridtools::clang;

stencil test{
        storage in, out;
        Do{
                vertical_region(k_start, k_end){
                        in =out;
                        out =in;
                }
        }
};
```

### Testing

Testing in https://github.com/MeteoSwiss-APN/clang-gridtools/pull/138

### CODE REVIEW

File: dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  ms->update(iir::NodeUpdateType::levelAndTreeAbove);
+
+  return ms;
+}
+
+/// @brief Collects AccessIDs from versioned fields in the IIR.
+struct CollectVersionedIDs : public iir::ASTVisitorForwarding {
+  const iir::StencilMetaInformation& metadata_;
+  std::set<int> versionedAccessIDs;
+
```
Comment (mroethlin) [inline dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp]: why not have this set private with a getter?

File: dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+          DAWN_LOG(WARNING) << "Cannot resolve race conditions with vertical dependencies. This "
+                               "will generate potentially dangerous code.";
+        }
+
+        // If it has a non-zero extent, we need to generate a multistage that
+        // fills all read intervals from the original field during the execution
+        // of the multistage. Currently, we create one for each interval, but
+        // these could later be merged into a single multistage.
+        const auto multiInterval = (*msiter)->computeReadAccessInterval(id);
+        for(auto interval : multiInterval.getIntervals()) {
```
Comment (mroethlin) [inline dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp]: const auto & ?

File: dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+  return domethod;
+}
+
+/// @brief Creates the stage in which assignment happens (where the temporary gets filled)
+static std::unique_ptr<iir::Stage>
+createAssignmentStage(int assignmentID, int assigneeID,
+                      std::shared_ptr<iir::StencilInstantiation> const& si,
+                      const iir::Interval& interval) {
+  // TODO: Use the IIRBuilder instead of doing this manually...
```
[the remark this answers is not in the document]
Comment (jdahm) [reply dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp]: :)

File: dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+#include <memory>
+
+namespace dawn {
+
+/// @brief Creates the assignment statement
+static std::shared_ptr<ast::Stmt>
+createAssignmentStatement(int assignmentID, int assigneeID,
+                          iir::StencilMetaInformation const& metadata) {
+  auto fa_assignee =
+      std::make_shared<ast::FieldAccessExpr>(metadata.getNameFromAccessID(assigneeID));
```
[the remark this answers is not in the document]
Comment (jdahm) [reply dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp]: Done!

File: dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+#include "dawn/Optimizer/OptimizerContext.h"
+#include "dawn/Support/Logging.h"
+
+#include <memory>
+
+namespace dawn {
+
+/// @brief Creates the assignment statement
+static std::shared_ptr<ast::Stmt>
+createAssignmentStatement(int assignmentID, int assigneeID,
```
[the remark this answers is not in the document]
Comment (jdahm) [reply dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp]: It seems like the function call that consumes it takes a `std::shared_ptr`.

File: dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp (hunk 1 above)
Comment (jdahm) [reply dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp]: I suppose, but that adds code bulk, and the example I based this on in `PassFieldVersioning.cpp` did not have one... What do you think?

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
The bugfix corrected some tests, which caused what were false positive to now fail. These tests are all related to boundary conditions, which are broken anyway, so the tests and the pass are temporarily disabled until we can redesign the boundary condition concepts.

**Comment by jdahm** [comment]
launch jenkins

File: dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+createAssignmentStage(int assignmentID, int assigneeID,
+                      std::shared_ptr<iir::StencilInstantiation> const& si,
+                      const iir::Interval& interval) {
+  // Create the do-method for the stage
+  auto doMethod = createDoMethod(assignmentID, assigneeID, si, interval);
+
+  // Create the stage and add the do method
+  iir::StencilMetaInformation& metadata = si->getMetaData();
+  auto assignmentStage = std::make_unique<iir::Stage>(metadata, si->nextUID());
+  assignmentStage->setExtents(iir::Extents{});
```
Comment (jdahm) [inline dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp]: If we do not want to run the metadata update pass, then we should get extents from the field instead of setting it to 0.

**Comment by jdahm** [comment]
launch jenkins

**Comment by twicki** [comment]
launch jenkins

**Comment by jdahm** [comment]
Before merging:
- address comment (update extents)
- remove `boundary_condition` temporarily from the frontend

**Comment by jdahm** [comment]
Depends on #138 in clang-gridtools.
